### PR TITLE
An attempt at an interface to csv text files containing numeric data.…

### PR DESCRIPTION
**Pull request by crcox**
… Output is formatted like an image

Fixes #{issue_number}.

### Description
Provide an interface to read numeric data from comma separated text. The text file should contain tabular data with one row per example. Features are represented as columns. Additional columns can differentiate subjects, domains, categories, and example indexes. It is currently required for the data to be "balanced": same number of examples for each subject, domain, and category. This is because the function returns a numpy array, similar to an `image`.

### Status
**Ready/Work in progress/Hold**

If you've been given access to pykale with role Triage or above, on the right (delete these after selection):

- Select **one most appropriate** label.
- **When** your pull request is **ready** for review, select a reviewer. Use the suggested one if unsure.

If you cannot see an option to select a reviewer/label, that means one maintainer will get notified upon your pull request and then select for you.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] In-line docstrings updated.
- [ ] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

File: kale/loaddata/numeric_acccess.py
Comment (haipinglu) [inline kale/loaddata/numeric_acccess.py:5]: `tabular` should be a more precise word for this module than `numeric`, i.e., `numeric_access` to `tabular_access`, and similarly for the rest

File: kale/loaddata/numeric_acccess.py
Comment (haipinglu) [inline kale/loaddata/numeric_acccess.py:23]: `print` should be avoided. See bullet #3 of https://github.com/pykale/pykale/blob/main/.github/CONTRIBUTING.md#coding-style

File: kale/loaddata/numeric_acccess.py
Comment (haipinglu) [inline kale/loaddata/numeric_acccess.py:1]: Some docstrings at the top could make it clear the purpose of this whole module

File: kale/loaddata/numeric_acccess.py
Comment (haipinglu) [inline kale/loaddata/numeric_acccess.py:27]: Some explanation on why `images` will help review and readability.

**Comment by haipinglu** [comment]
@crcox Many thanks for your contribution!
- Note that we are doing linting so to pass linting, you need to follow https://github.com/pykale/pykale/blob/main/.github/CONTRIBUTING.md#before-pull-requests-pre-commit-hooks to set up pre-commit checks so that code committed are consistent with existing code.
- We are following some common software engineering practices so if you could go through our guidelines for contributors https://github.com/pykale/pykale/blob/main/.github/CONTRIBUTING.md#medium-involvements-contributors, the PR/merging will become smoother. 

Feel free to ask if anything is unclear to you. I also invited you with write access just now.

File: kale/loaddata/numeric_acccess.py
Comment (haipinglu) [reply kale/loaddata/numeric_acccess.py:5]: @sz144 What do you think about `tabular` vs `numeric`?

Comment (sz144) [reply kale/loaddata/numeric_acccess.py:5]: Tabular is more general, while numeric is more precise. I prefer `numeric` if it will not support other data type in future.

Comment (haipinglu) [reply kale/loaddata/numeric_acccess.py:5]: We do need to support EHR data via `Tabular`, e.g. what Mwiza did on our ABIDE data and what Mohammod is working on our cardiac data.

File: kale/loaddata/numeric_acccess.py
Comment (crcox) [reply kale/loaddata/numeric_acccess.py:23]: @haipinglu Thank you for this guidance. I actually want to raise an exception, but I couldn't remember how :) I will fix it.

**Comment by crcox** [comment]
I am having trouble setting up my development environment so that the pre-commit routine works. I have made revisions that address some of the recommendations (and fixed some broken code), but I am failing to commit because of errors, I think:

```
> git -c user.useConfigOnly=true commit --quiet --allow-empty-message --file -
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://gitlab.com/pycqa/flake8.
[INFO] Initializing environment for https://gitlab.com/pycqa/flake8:flake8-print.
[INFO] Initializing environment for https://github.com/psf/black.
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-isort.
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://gitlab.com/pycqa/flake8.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/psf/black.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
Check for added large files..............................................Passed
check BOM - deprecated: use fix-byte-order-marker........................Passed
Check for case conflicts.................................................Passed
Check for merge conflicts................................................Passed
Fix End of Files.........................................................Passed
Forbid new submodules....................................................Passed
Mixed line ending........................................................Passed
Trim Trailing Whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing kale/loaddata/tabular_access.py

Debug Statements (Python)................................................Passed
Check Yaml...........................................(no files to check)Skipped
Fix requirements.txt.................................(no files to check)Skipped
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

kale/loaddata/tabular_access.py:10:1: E302 expected 2 blank lines, found 1
kale/loaddata/tabular_access.py:11:41: E231 missing whitespace after ':'
kale/loaddata/tabular_access.py:12:50: E231 missing whitespace after ','
kale/loaddata/tabular_access.py:16:30: E231 missing whitespace after ':'
kale/loaddata/tabular_access.py:16:45: E231 missing whitespace after ':'
kale/loaddata/tabular_access.py:33:31: E231 missing whitespace after ':'
kale/loaddata/tabular_access.py:33:39: E231 missing whitespace after ','

black....................................................................Failed
- hook id: black
- exit code: 1

Traceback (most recent call last):
  File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/bin/black", line 8, in <module>
    sys.exit(patched_main())
  File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/black.py", line 4134, in patched_main
    patch_click()
  File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/black.py", line 4123, in patch_click
    from click import _unicodefun  # type: ignore
ImportError: cannot import name '_unicodefun' from 'click' (/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/click/__init__.py)

isort....................................................................Failed
- hook id: isort
- exit code: 1

/home/chriscox/.pyenv/versions/3.7.3/bin/python: No module named isort

mypy.....................................................................Passed
```

**Comment by haipinglu** [comment]
> I am having trouble setting up my development environment so that the pre-commit routine works. I have made revisions that address some of the recommendations (and fixed some broken code), but I am failing to commit because of errors, I think.
> 
> Edit: @sz144 Could you possibly help me get past this?
> 
> ```
> > git -c user.useConfigOnly=true commit --quiet --allow-empty-message --file -
> [INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
> [INFO] Initializing environment for https://gitlab.com/pycqa/flake8.
> [INFO] Initializing environment for https://gitlab.com/pycqa/flake8:flake8-print.
> [INFO] Initializing environment for https://github.com/psf/black.
> [INFO] Initializing environment for https://github.com/pre-commit/mirrors-isort.
> [INFO] Initializing environment for https://github.com/pre-commit/mirrors-mypy.
> [INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
> [INFO] Once installed this environment will be reused.
> [INFO] This may take a few minutes...
> [INFO] Installing environment for https://gitlab.com/pycqa/flake8.
> [INFO] Once installed this environment will be reused.
> [INFO] This may take a few minutes...
> [INFO] Installing environment for https://github.com/psf/black.
> [INFO] Once installed this environment will be reused.
> [INFO] This may take a few minutes...
> [INFO] Installing environment for https://github.com/pre-commit/mirrors-mypy.
> [INFO] Once installed this environment will be reused.
> [INFO] This may take a few minutes...
> Check for added large files..............................................Passed
> check BOM - deprecated: use fix-byte-order-marker........................Passed
> Check for case conflicts.................................................Passed
> Check for merge conflicts................................................Passed
> Fix End of Files.........................................................Passed
> Forbid new submodules....................................................Passed
> Mixed line ending........................................................Passed
> Trim Trailing Whitespace.................................................Failed
> - hook id: trailing-whitespace
> - exit code: 1
> - files were modified by this hook
> 
> Fixing kale/loaddata/tabular_access.py
> 
> Debug Statements (Python)................................................Passed
> Check Yaml...........................................(no files to check)Skipped
> Fix requirements.txt.................................(no files to check)Skipped
> flake8...................................................................Failed
> - hook id: flake8
> - exit code: 1
> 
> kale/loaddata/tabular_access.py:10:1: E302 expected 2 blank lines, found 1
> kale/loaddata/tabular_access.py:11:41: E231 missing whitespace after ':'
> kale/loaddata/tabular_access.py:12:50: E231 missing whitespace after ','
> kale/loaddata/tabular_access.py:16:30: E231 missing whitespace after ':'
> kale/loaddata/tabular_access.py:16:45: E231 missing whitespace after ':'
> kale/loaddata/tabular_access.py:33:31: E231 missing whitespace after ':'
> kale/loaddata/tabular_access.py:33:39: E231 missing whitespace after ','
> 
> black....................................................................Failed
> - hook id: black
> - exit code: 1
> 
> Traceback (most recent call last):
>   File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/bin/black", line 8, in <module>
>     sys.exit(patched_main())
>   File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/black.py", line 4134, in patched_main
>     patch_click()
>   File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/black.py", line 4123, in patch_click
>     from click import _unicodefun  # type: ignore
> ImportError: cannot import name '_unicodefun' from 'click' (/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/click/__init__.py)
> 
> isort....................................................................Failed
> - hook id: isort
> - exit code: 1
> 
> /home/chriscox/.pyenv/versions/3.7.3/bin/python: No module named isort
> 
> mypy.....................................................................Passed
> ```

@sz144 Did you miss this message/update?

@crcox Chris: Sorry for late response from us. What IDE are you using? Have you followed https://github.com/pykale/pykale/blob/main/.github/CONTRIBUTING.md#before-pull-requests-pre-commit-hooks to install pre-commit and the hooks from the root directory via
```sh
pip install pre-commit
pre-commit install
```
If you do, it should install packages needed automatically for you. From your posted message, it seems that you may not have done that (e.g. it says isort not installed). 

If successfully installed, then every time you commit, it will run the check and automatically fix what it can fix and report the remaining to you. Typically, flake8 and black errors need to be fixed manually but usually we install plug-ins so those issues can be automatically detected (like a typo) and fixed right away. Once all being fixed, you need to commit the changes again to pass the pre-commit checks.

It may sound complicated but once done, it is beneficial. We can use our May 4th meeting or set up a meeting with you before that to help you set up. We highly recommend you to use vscode (me) or pycharm (Shuo) to commit code.

I also hope to learn how to make our contributing guidelines clearer for new comers like you. We need to make our guidelines clearer. Thank you very much!

**Comment by crcox** [comment]
Thank you! I am using VS Code. I think I understand my problem, but not exactly how to fix it. I installed all the pykale and pre-commit dependencies in a `venv`, but I put the `venv` in a location outside the project folder. So even though I activate that environment in one terminal, I think pre-commit is firing up a fresh terminal where that environment is not active.

Now I have installed pre-commit in a more standard location, and am using Python 3.8.12 as I originally intended. Now I am getting this error:

```
black....................................................................Failed
- hook id: black
- exit code: 1

Traceback (most recent call last):
  File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/bin/black", line 8, in <module>
    sys.exit(patched_main())
  File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/black.py", line 4134, in patched_main
    patch_click()
  File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/black.py", line 4123, in patch_click
    from click import _unicodefun  # type: ignore
ImportError: cannot import name '_unicodefun' from 'click' (/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/click/__init__.py)
```

That seems related to this: https://github.com/dask/distributed/issues/6013

So I installed `click==8.0.2` as shown in the MWE, but that did not alleviate the issue.

These are the packages installed in the relevant environment:

```
[chriscox@relearn0 pykale]$ python --version
Python 3.8.12
[chriscox@relearn0 pykale]$ pip list
Package               Version
--------------------- -------
astroid               2.11.2
cfgv                  3.3.1
click                 8.0.2
dill                  0.3.4
distlib               0.3.4
filelock              3.6.0
identify              2.4.12
isort                 5.10.1
lazy-object-proxy     1.7.1
matlabengineforpython R2021a
mccabe                0.7.0
nodeenv               1.6.0
pip                   22.0.4
platformdirs          2.5.1
pre-commit            2.18.1
pylint                2.13.5
PyYAML                6.0
setuptools            62.1.0
six                   1.16.0
toml                  0.10.2
tomli                 2.0.1
typing_extensions     4.1.1
virtualenv            20.14.1
wheel                 0.37.1
wrapt                 1.14.0
```

**Comment by haipinglu** [comment]
> Thank you! I am using VS Code. I think I understand my problem, but not exactly how to fix it. I installed all the pykale and pre-commit dependencies in a `venv`, but I put the `venv` in a location outside the project folder. So even though I activate that environment in one terminal, I think pre-commit is firing up a fresh terminal where that environment is not active.
> 
> Now I have installed pre-commit in a more standard location, and am using Python 3.8.12 as I originally intended. Now I am getting this error:
> 
> ```
> black....................................................................Failed
> - hook id: black
> - exit code: 1
> 
> Traceback (most recent call last):
>   File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/bin/black", line 8, in <module>
>     sys.exit(patched_main())
>   File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/black.py", line 4134, in patched_main
>     patch_click()
>   File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/black.py", line 4123, in patch_click
>     from click import _unicodefun  # type: ignore
> ImportError: cannot import name '_unicodefun' from 'click' (/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/click/__init__.py)
> ```
> 
Great to see that only one problem remaining. I would suggest to test `black` individually, as suggest at https://github.com/pykale/pykale/blob/main/.github/CONTRIBUTING.md#manual-checks-and-fixes-be-careful to manually check what the problem is. It could give you more info and help you track down the problem.

> That seems related to this: [dask/distributed#6013](https://github.com/dask/distributed/issues/6013)
> 
> So I installed `click==8.0.2` as shown in the MWE, but that did not alleviate the issue.
> 
> These are the packages installed in the relevant environment:
> 
> ```
> [chriscox@relearn0 pykale]$ python --version
> Python 3.8.12
> [chriscox@relearn0 pykale]$ pip list
> Package               Version
> --------------------- -------
> astroid               2.11.2
Where is `black`?
> cfgv                  3.3.1
> click                 8.0.2
> dill                  0.3.4
> distlib               0.3.4
> filelock              3.6.0
> identify              2.4.12
> isort                 5.10.1
> lazy-object-proxy     1.7.1
> matlabengineforpython R2021a
> mccabe                0.7.0
> nodeenv               1.6.0
> pip                   22.0.4
> platformdirs          2.5.1
> pre-commit            2.18.1
> pylint                2.13.5
> PyYAML                6.0
> setuptools            62.1.0
> six                   1.16.0
> toml                  0.10.2
> tomli                 2.0.1
> typing_extensions     4.1.1
> virtualenv            20.14.1
> wheel                 0.37.1
> wrapt                 1.14.0
> ```
Did you have `black` installed (properly)? `black` is not listed in the list above so this may be the cause.

**Comment by haipinglu** [comment]
> @haipinglu I installed back and ran it independently on this source file. I get:
> 
> ```
> [chriscox@relearn0 pykale]$ black kale/loaddata/tabular_access.py 
> reformatted kale/loaddata/tabular_access.py
> 
> All done! ✨ 🍰 ✨
> 1 file reformatted.
> ```
> 
> When I try to commit, I get:
> 
> ```
> [2022-04-25T15:39:18.206Z] > git symbolic-ref --short HEAD [2ms]
> [2022-04-25T15:39:18.215Z] > git for-each-ref --format=%(refname)%00%(upstream:short)%00%(objectname)%00%(upstream:track) refs/heads/numeric_text_access refs/remotes/numeric_text_access [3ms]
> [2022-04-25T15:39:18.232Z] > git for-each-ref --sort -committerdate --format %(refname) %(objectname) %(*objectname) [9ms]
> [2022-04-25T15:39:18.233Z] > git remote --verbose [2ms]
> [2022-04-25T15:39:18.244Z] > git config --get commit.template [2ms]
> [2022-04-25T15:39:30.439Z] git.stage 1
> [2022-04-25T15:39:30.440Z] git.stage.scmResources 1
> [2022-04-25T15:39:30.452Z] > git add -A -- /home/chriscox/GitHub/forks/pykale/kale/loaddata/tabular_access.py [3ms]
> [2022-04-25T15:39:30.472Z] > git symbolic-ref --short HEAD [1ms]
> [2022-04-25T15:39:30.480Z] > git for-each-ref --format=%(refname)%00%(upstream:short)%00%(objectname)%00%(upstream:track) refs/heads/numeric_text_access refs/remotes/numeric_text_access [2ms]
> [2022-04-25T15:39:30.492Z] > git for-each-ref --sort -committerdate --format %(refname) %(objectname) %(*objectname) [6ms]
> [2022-04-25T15:39:30.493Z] > git remote --verbose [2ms]
> [2022-04-25T15:39:30.504Z] > git config --get commit.template [2ms]
> [2022-04-25T15:39:31.571Z] > git symbolic-ref --short HEAD [2ms]
> [2022-04-25T15:39:31.579Z] > git for-each-ref --format=%(refname)%00%(upstream:short)%00%(objectname)%00%(upstream:track) refs/heads/numeric_text_access refs/remotes/numeric_text_access [2ms]
> [2022-04-25T15:39:31.592Z] > git for-each-ref --sort -committerdate --format %(refname) %(objectname) %(*objectname) [7ms]
> [2022-04-25T15:39:31.593Z] > git remote --verbose [2ms]
> [2022-04-25T15:39:31.607Z] > git config --get commit.template [2ms]
> [2022-04-25T15:39:31.672Z] > git ls-files --stage -- /home/chriscox/GitHub/forks/pykale/kale/loaddata/tabular_access.py [3ms]
> [2022-04-25T15:39:31.680Z] > git cat-file -s 9d0dc4d764beb4542ae4d648316cbc1923316418 [2ms]
> [2022-04-25T15:39:40.387Z] > git ls-files --stage -- /home/chriscox/GitHub/forks/pykale/kale/loaddata/tabular_access.py [2ms]
> [2022-04-25T15:39:40.399Z] > git cat-file -s 9d0dc4d764beb4542ae4d648316cbc1923316418 [2ms]
> [2022-04-25T15:39:40.777Z] > git -c user.useConfigOnly=true commit --quiet --allow-empty-message --file - [1572ms]
> [2022-04-25T15:39:40.777Z] Check for added large files..............................................Passed
> check BOM - deprecated: use fix-byte-order-marker........................Passed
> Check for case conflicts.................................................Passed
> Check for merge conflicts................................................Passed
> Fix End of Files.........................................................Passed
> Forbid new submodules....................................................Passed
> Mixed line ending........................................................Passed
> Trim Trailing Whitespace.................................................Passed
> Debug Statements (Python)................................................Passed
> Check Yaml...........................................(no files to check)Skipped
> Fix requirements.txt.................................(no files to check)Skipped
> flake8...................................................................Passed
> black....................................................................Failed
> - hook id: black
> - exit code: 1
> 
> Traceback (most recent call last):
>   File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/bin/black", line 8, in <module>
>     sys.exit(patched_main())
>   File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/black.py", line 4134, in patched_main
>     patch_click()
>   File "/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/black.py", line 4123, in patch_click
>     from click import _unicodefun  # type: ignore
> ImportError: cannot import name '_unicodefun' from 'click' (/home/chriscox/.cache/pre-commit/repo3rh1wov5/py_env-python3/lib/python3.8/site-packages/click/__init__.py)
> 
> isort....................................................................Passed
> mypy.....................................................................Passed
> [2022-04-25T15:39:40.786Z] > git config --get-all user.name [2ms]
> [2022-04-25T15:39:40.794Z] > git config --get-all user.email [2ms]
> ```
> 
> This seems like a lot of hoops to standardize white-space, or whatever this all does!

@crcox 
- Can you email the file to me and let me create a new branch to commit locally to see whether I will get the same error? (I won't merge).
- Alternatively, create a sandbox branch (for testing, not for merging) and do some very minor changes (e.g. change of doc, readme) to see whether you can commit and PR.
- BTW, I invited you earlier to this repo but you did not accept and it expired. Would you like me to re-invite  you and join this repo to collaborate directly (preferred and more efficient for debug etc)?

**Comment by crcox** [comment]
I figured it out---the `pre-commit` packages are installed into a new python environment that I was not aware of, and in that environment the wrong version of `click` was being installed by default. Once I rolled it back to version `8.0.2`, `pre-commit run` completed without error. I pushed to the same branch and tried to open a new pull request, but maybe it will just update this one that exists?

With respect to becoming an contributor on the main-line repository... I don't know if I want to do that, because it is probably best if all my changes go through the pull request code review, since I am not as experienced as the rest of the team. I don't really understand the relative costs and benefits of becoming a collaborator.

**Comment by haipinglu** [comment]
> I figured it out---the `pre-commit` packages are installed into a new python environment that I was not aware of, and in that environment the wrong version of `click` was being installed by default. Once I rolled it back to version `8.0.2`, `pre-commit run` completed without error. I pushed to the same branch and tried to open a new pull request, but maybe it will just update this one that exists?

@crcox Great to see you found and fixed the problem. Could you help make the [respective contribution guideline](https://github.com/pykale/pykale/blob/main/.github/CONTRIBUTING.md#before-pull-requests-pre-commit-hooks) clearer so that if someone else has the same/similar problem could learn from your experience to identify and fix the problem more quickly? Thanks. It could be a separate PR, preferably, (e.g. as a contributor, see reply below about a clarification).

> 
> With respect to becoming an contributor on the main-line repository... I don't know if I want to do that, because it is probably best if all my changes go through the pull request code review, since I am not as experienced as the rest of the team. I don't really understand the relative costs and benefits of becoming a collaborator.

@crcox I did not explain earlier. Our main branch is protected so that all PR to main MUST have at least one approval review and pass all the checks to be able to merge, including my (admin) PR. Otherwise, PR is blocked. Nobody, neither me nor you, can merge a PR without an approval review. As a contributor, you create branches directly and that will be easier for others to be aware of what you are working on (e.g. to avoid duplicated efforts, such as fixing a bug) and when there is a problem, it is easier to work collaborative to fix the problems together. If you are still unsure, I could explain more to you on 4th May.

**Comment by haipinglu** [comment]
@crcox Could you edit the title of this PR following the style of others https://github.com/pykale/pykale/blob/main/.github/CONTRIBUTING.md#before-pull-requests-pre-commit-hooks? Typically it should start with a verb to describe the PR (actions). This title will automatically be summarized in the [CHANGELOG](https://github.com/pykale/pykale/blob/main/.github/CHANGELOG.md) in the next release (see the first subitem of the third item in https://github.com/pykale/pykale/blob/main/.github/CONTRIBUTING.md#branch-fork-and-pull). 

Therefore, if the title is in a consistent format, this could save the need to manually edit the CHANGELOG in the next release. This applies to all your future PRs. Thanks.

File: kale/loaddata/tabular_access.py
Comment (haipinglu) [inline kale/loaddata/tabular_access.py:2]: I suggest to change "disk" to "a path".

File: kale/loaddata/tabular_access.py
Comment (haipinglu) [inline kale/loaddata/tabular_access.py:11]: Some docstring here to explain a bit to help reading

**Comment by sz144** [comment]
@crcox Great to see the `pre-commit` finally works. Because this is a new API, could you please update the [source of documentation](https://github.com/crcox/pykale/blob/numeric_text_access/docs/source/kale.loaddata.rst). We will try to make it automatic in the future, but now we have to do this manually.

File: kale/loaddata/tabular_access.py
Comment (haipinglu) [inline kale/loaddata/tabular_access.py:23]: At https://github.com/pykale/pykale/blob/main/setup.cfg#L8, you can see that we use max-line-length of 120, which is more common nowadays (e.g. in PyTorch). You are probably using the default value 80(?) in black formatting. I suggest you to change the setting to 120 to be consistent in our coding style.

File: kale/loaddata/tabular_access.py
Comment (haipinglu) [inline kale/loaddata/tabular_access.py:22]: At https://github.com/pykale/pykale/blob/main/setup.cfg#L8, you can see that we use max-line-length of 120, which is more common nowadays (e.g. in PyTorch). You are probably using the default value 80(?) in black formatting. I suggest you to change the setting to 120 to be consistent in our coding style.

File: kale/loaddata/tabular_access.py
Comment (haipinglu) [inline kale/loaddata/tabular_access.py:37]: remove a redundant "an"

File: kale/loaddata/tabular_access.py
Comment (haipinglu) [inline kale/loaddata/tabular_access.py:40]: This (future work) should be captured, e.g. as a card in a [project](https://github.com/pykale/pykale/projects), or a discussion or issue.

File: kale/loaddata/tabular_access.py
Comment (haipinglu) [inline kale/loaddata/tabular_access.py:43]: Will "arrays" be better and more accurate than "images" (max 4D)?

**Review by haipinglu — CHANGES_REQUESTED** [review]
Need to add documentation. You are adding a Second-level module so you could follow the instruction Step 2 at https://github.com/pykale/pykale/tree/main/docs#workflow, keeping the modules sorted. We have automate the steps so Step 3 is only needed if you want to verify locally. You may skip step 3.

Also need to add a test to verify that it is working. See https://github.com/pykale/pykale/tree/main/tests#test-guidelines  The easiest way to start is by looking at other examples. You could generate some toy data for testing.

If these comments seem complicated and a lot of work for you, we could coach you through in our 4th May meeting or a meeting separately. Once you do it once, it will become relatively easy. These are software engineering practices widely adopted for mature/robust libraries. 

Your experience can also help us improve the guideline to engage more contributors in future. Many thanks.

**Comment by haipinglu** [comment]
@crcox Also, please add a label so that it can be properly categorised (e.g. in CHANGELOG).

**Comment by haipinglu** [comment]
We have a clearer understanding of the problem and this PR is no longer up to date.

**Pull request closed without merging**
